### DISCUSSION TITLE
Auto deletion of inactive durable consumers

### DISCUSSION THREAD

**Discussion by beatrausch** [nats-io/nats.go#924] [Q&A] [answered]
Hi, 

is there the possibility to configure nats to delete a durable consumer after a configurable amount of time of inactivity?

Regards

**Comment by bruth** [comment, marked as answer]
@beatrausch Can you define what you mean by "inactivity" and/or describe the use case in more detail? A consumer is a _view_ of a stream and then one or more subscriptions are created (and bound) to that consumer to actually do the message handling.

**Reply by beatrausch** [reply]
@bruth  Let’s say there is an app A for which a consumer A exists and this app is active over 3 months meaning there are from time-to-time subscriptions to its consumer A. After 3 months there are no more subscription to consumer A. The question is, if nats provide a mechanism to automatically remove this kind of “unused” consumer after a configurable amount of time of inactivity?

**Reply by bruth** [reply]
The short answer is no, there is no auto-deletion of consumers. However, consumer state is fairly minimal. Meaning even if consumers are around longer than they need to be, but are not active, there is only a small bit of state associated with them.

So then the question is how to detect and then clean up consumers. There is a `Last` (last active) field on [`SequenceInfo`](https://pkg.go.dev/github.com/nats-io/nats.go#SequenceInfo) on [`ConsumerInfo.Delivered`](https://pkg.go.dev/github.com/nats-io/nats.go#SequenceInfo) which would indicated the last time a message was delivered to any subscriber.

You could have a small program that loops through all consumers and checks that "last active" field to see if it passed the threshold of inactivity to be cleaned up.

**Reply by derekcollison** [reply]
Ephemeral consumers that are not active will be auto cleaned up.

For pull based (this is new so most clients not supporting it yet) this means no pull requests in the Inactivity time window (You can now set this as well).

For push it means the consumer is not bound for the same window.

**Reply by derekcollison** [reply]
https://github.com/nats-io/nats-server/blob/main/server/consumer.go#L78

**Reply by grillorafael** [reply]
hi @derekcollison, is this still a thing ? This link points to another place in the file now.

**Reply by ripienaar** [reply]
@grillorafael yes, all consumers have inactive timeout setting that can clean them up

**Reply by grillorafael** [reply]
So does that mean I can use `InactiveThreshold` on Durable consumers?

**Reply by Jarema** [reply]
That is correct.
